File: internal/domain/asset/asset.go
package asset

import (
	"context"
	"time"
)

// AssetBatch 入库批次(公司采购行为,资产经此归属公司)。
type AssetBatch struct {
	ID            int64  `json:"id"`
	LegalEntityID int64  `json:"legalEntityId"`
	Code          string `json:"code"` // 批次编码,如 RK-202607-01
	Name          string `json:"name"`
}

// Tag 电子标签(公司库存,预绑定后才指向资产)。
type Tag struct {
	TagID         int64  `json:"tagId"`
	LegalEntityID int64  `json:"legalEntityId"`
	TagNo         string `json:"tagNo"`        // 标签编号
	EpcCode       string `json:"epcCode"`      // EPC 码
	Band          string `json:"band"`         // 频段,如 UHF
	BoundAssetID  int64  `json:"boundAssetId"` // 0=未绑定
	Status        string `json:"status"`       // UNBOUND/BOUND/DISABLED
	Battery       string `json:"battery"`
}

// TagEvent 标签绑定事件(append-only 审计流,P1-T2):BIND/UNBIND/RECYCLE,
// 形态依据 R2 调研(Snipe-IT action_logs / bk-cmdb cc_AuditLog)。
type TagEvent struct {
	ID             int64          `json:"id"`
	EventID        string         `json:"eventId"`
	TagID          int64          `json:"tagId"`
	AssetID        int64          `json:"assetId"`
	Action         string         `json:"action"`
	ActorAccountID int64          `json:"actorAccountId"`
	Detail         string         `json:"detail"`
	Changed        map[string]any `json:"changed,omitempty"`
	CreatedAt      time.Time      `json:"createdAt"`
}

// AssetModel 资产型号字典(P1-T3,R3 调研:NetBox DeviceType/GLPI models 同款形态):
// UNIQUE(vendor,model,category,part_number) 防重;is_active 停用不物理删(引用保护)。
type AssetModel struct {
	ID         int64          `json:"id"`
	Vendor     string         `json:"vendor"`
	Model      string         `json:"model"`
	Category   string         `json:"category"`
	PartNumber string         `json:"partNumber"`
	Spec       map[string]any `json:"spec,omitempty"`
	IsActive   bool           `json:"isActive"`
	CreatedAt  time.Time      `json:"createdAt"`
}

// Asset 资产台账(光猫/ONU 等装维物资的全生命周期)。
type Asset struct {
	AssetID         int64  `json:"assetId"`
	AssetCode       string `json:"assetCode"` // 资产编码,如 A-20260001
	BatchID         int64  `json:"batchId"`
	LegalEntityID   int64  `json:"legalEntityId"` // 企业归属快照
	LegalEntityName string `json:"legalEntityName"`
	TagID           int64  `json:"tagId"`     // 0=未绑定
	AddressID       int64  `json:"addressId"` // 0=未部署
	RegionID        int64  `json:"regionId"`  // 0=未部署
	RegionName      string `json:"regionName"`
	Type            string `json:"type"`    // 光猫/ONU/路由器(展示冗余;权威=model_id→asset_models.category)
	ModelID         int64  `json:"modelId"` // 0=未挂型号(P1-T3)
	Status          string `json:"status"`  // IN_STOCK/DEPLOYED/MAINTENANCE/SCRAPPED
}

// AssetLifecycle 资产状态轨迹(每次状态/位置变更一行,历史不随当前状态漂移)。
type AssetLifecycle struct {
	ID          int64     `json:"id"`
	AssetID     int64     `json:"assetId"`
	Status      string    `json:"status"`
	AddressID   int64     `json:"addressId"` // 0=空
	AddressName string    `json:"addressName"`
	WorkerID    int64     `json:"workerId"` // 0=空
	WorkerName  string    `json:"workerName"`
	ChangedAt   time.Time `json:"changedAt"`
}

// Replacement 换新单(故障资产换新流程)。
// 状态机: PENDING --Assign--> DOING --Complete--> DONE/FAILED(adopted note
// 2026-08-27-replacement-ticket-flow);终态不可再流转,重做走新单。
type Replacement struct {
	ID              int64      `json:"id"`
	ReplacementNo   string     `json:"replacementNo"`
	AssetID         int64      `json:"assetId"`
	LegalEntityID   int64      `json:"legalEntityId"`
	LegalEntityName string     `json:"legalEntityName"`
	Reason          string     `json:"reason"`
	Priority        string     `json:"priority"` // HIGH/MEDIUM/LOW
	Status          string     `json:"status"`   // PENDING/DOING/DONE/FAILED
	WorkerID        int64      `json:"workerId"` // 0=未派
	WorkerName      string     `json:"workerName"`
	FinishedAt      *time.Time `json:"finishedAt,omitempty"` // nil=未完成
}

// AssetAssignment 资产持有台账(每次领用/部署/归还的时间段,历史不随当前值漂移)。
type AssetAssignment struct {
	ID                int64      `json:"id"`
	AssetID           int64      `json:"assetId"`
	WorkerID          int64      `json:"workerId"` // 0=空
	WorkerName        string     `json:"workerName"`
	AddressID         int64      `json:"addressId"` // 0=空
	AddressName       string     `json:"addressName"`
	Reason            string     `json:"reason"`
	OperatorAccountID int64      `json:"operatorAccountId"` // 0=空
	EffectiveFrom     time.Time  `json:"effectiveFrom"`
	EffectiveTo       *time.Time `json:"effectiveTo,omitempty"` // nil=至今
}

// Stocktake 盘点任务(按区域盘点资产,输出差异)。
type Stocktake struct {
	ID            int64  `json:"id"`
	LegalEntityID int64  `json:"legalEntityId"`
	Scope         string `json:"scope"`
	Progress      int16  `json:"progress"` // 0~100,实扫/计划快照行
	DiffCount     int32  `json:"diffCount"`
	Status        string `json:"status"` // DOING/DONE
}

// StocktakeItem 盘点差异明细(建单冻结快照,扫码回填,逐条处置)。
type StocktakeItem struct {
	ID             int64      `json:"id"`
	TaskID         int64      `json:"taskId"`
	AssetID        int64      `json:"assetId"`
	ExpectedStatus string     `json:"expectedStatus"` // ""=计划外(EXTRA 行)
	ScannedStatus  string     `json:"scannedStatus"`  // ""=未扫
	ScannedAt      *time.Time `json:"scannedAt,omitempty"`
	Kind           string     `json:"kind"`       // PENDING/OK/MISMATCH/MISSING/EXTRA
	Resolution     string     `json:"resolution"` // OPEN/CONFIRMED/FIXED/ESCALATED
	HandledBy      int64      `json:"handledBy"`  // 0=未处置
	HandledAt      *time.Time `json:"handledAt,omitempty"`
	Note           string     `json:"note"`
}

// AssetService 资产域服务口(阶段3):入库批次/电子标签/资产台账/状态轨迹/换新/盘点。
type AssetService interface {
	ListBatches(ctx context.Context) ([]AssetBatch, error)
	CreateBatch(ctx context.Context, b AssetBatch) (int64, error)
	ListTags(ctx context.Context) ([]Tag, error)
	CreateTag(ctx context.Context, t Tag) (int64, error)
	ListAssets(ctx context.Context) ([]Asset, error)
	CreateAsset(ctx context.Context, a Asset) (int64, error)
	GetAsset(ctx context.Context, id int64) (*Asset, error)
	// UpdateAsset 受限编辑(P2-W1-T1):仅 类型/型号/标签/批次 四键;标签换绑同一
	// 事务写 UNBIND+BIND(冲突 ErrBindingConflict 整单回滚);批次仅 IN_STOCK 可改
	// (ErrBatchNotEditable)并同步企业归属快照;四键无变化幂等成功。
	UpdateAsset(ctx context.Context, assetID int64, in AssetUpdate, actorAccountID int64) error
	// DeleteAsset 守卫删除(P2-W1-T1):仅 IN_STOCK 且无标签绑定/持有台账/换新单/
	// 盘点明细/四码关联引用可物理删除;命中引用 ErrAssetReferenced(message 全量
	// 列阻断项);SCRAPPED ErrAssetScrapped 拒硬删。返回资产编码供审计载荷。
	DeleteAsset(ctx context.Context, assetID int64) (string, error)

	ListLifecycles(ctx context.Context, assetID int64) ([]AssetLifecycle, error)
	AppendLifecycle(ctx context.Context, l AssetLifecycle) (int64, error)
	// SetAssetStatus 直改资产当前状态(换新完成联动);未命中返回 ErrNotFound。
	SetAssetStatus(ctx context.Context, assetID int64, status string) error
	ListReplacements(ctx context.Context) ([]Replacement, error)
	CreateReplacement(ctx context.Context, r Replacement) (int64, error)
	// GetReplacement 按 id 查换新单;未命中返回 ErrNotFound。
	GetReplacement(ctx context.Context, id int64) (*Replacement, error)
	// ListReplacementsByWorker 列出师傅名下的换新单(师傅端任务列表)。
	ListReplacementsByWorker(ctx context.Context, workerID int64) ([]Replacement, error)
	// AssignReplacement 派单:回填师傅快照并 PENDING→DOING;
	// 单不存在返回 ErrNotFound,状态非 PENDING 返回 ErrInvalidTransition。
	AssignReplacement(ctx context.Context, id, workerID int64, workerName string) (*Replacement, error)
	// CompleteReplacement 完成/失败:DOING→DONE|FAILED 并回填 finished_at;
	// 状态非 DOING 返回 ErrInvalidTransition。result 仅接受 DONE/FAILED。
	CompleteReplacement(ctx context.Context, id int64, result string) (*Replacement, error)
	ListStocktakes(ctx context.Context) ([]Stocktake, error)
	CreateStocktake(ctx context.Context, s Stocktake) (int64, error)
	// ListStocktakeItems 盘点差异明细清单。
	ListStocktakeItems(ctx context.Context, taskID int64) ([]StocktakeItem, error)
	// ScanStocktake 回填一次扫码,返回(明细 id, kind)。
	ScanStocktake(ctx context.Context, taskID, assetID int64, scannedStatus string) (int64, string, error)
	// HandleStocktakeItem 逐条处置差异(action: CONFIRM/FIX/ESCALATE)。
	HandleStocktakeItem(ctx context.Context, taskID, itemID int64, action, note string, accountID int64) error
	// HandleStocktakeDiff 关单:未处置差异非 0 时拒绝,全处置完任务置 DONE。
	HandleStocktakeDiff(ctx context.Context, taskID int64) error

	ListAssignments(ctx context.Context, assetID int64) ([]AssetAssignment, error)
	AssignAsset(ctx context.Context, a AssetAssignment) (int64, error)

	// UnbindTag 解绑标签(P1-T2):置 bound_asset_id=NULL+status=UNBOUND 并写 UNBIND 事件;
	// expectedAssetID>0 时校验当前绑定一致;未绑定/预期不符返回 ErrTagUnbound/ErrBindingConflict。
	UnbindTag(ctx context.Context, tagID, expectedAssetID, actorAccountID int64, detail string) error
	// DisableTag 停用标签(P2-W2-T1):仅 UNBOUND 可停用(BOUND 必须先解绑,
	// ErrBindingConflict 40900);已 DISABLED 幂等成功;未命中 ErrNotFound。
	DisableTag(ctx context.Context, tagID int64, reason string) error
	// EnableTag 启用标签(P2-W2-T1):仅对 DISABLED 生效(DISABLED → UNBOUND);
	// UNBOUND/BOUND 幂等 no-op 成功;未命中 ErrNotFound。
	EnableTag(ctx context.Context, tagID int64) error
	// ListTagEvents 标签事件流(P2-W2-T1):append-only 审计流只读回放,按时间倒序。
	ListTagEvents(ctx context.Context, tagID int64) ([]TagEvent, error)
	// ScrapAsset 报废资产(P1-T2):任意非终态 → SCRAPPED(终态幂等 no-op),强制解绑标签写
	// RECYCLE 事件(软回收禁硬删),轨迹落行;同一事务,失败整单回滚。
	ScrapAsset(ctx context.Context, assetID, actorAccountID int64, reason string) error

	// ListModels 型号字典(含停用,管理端下拉与列表)。
	ListModels(ctx context.Context) ([]AssetModel, error)
	// CreateModel 建型号:UNIQUE(vendor,model,category,part_number) 冲突返回 ErrModelExists。
	CreateModel(ctx context.Context, m AssetModel) (int64, error)
	// UpdateModel 编辑型号(P2-W2-T1):厂商/型号名/类别/料号/规格可改;唯一冲突
	// ErrModelExists(40900);停用型号拒绝编辑 ErrModelInactive(40900,先启用)。
	UpdateModel(ctx context.Context, id int64, m AssetModel) error
}

File: server-ts/src/entities/asset.ts
// L4 资产/标签/批次 + L5 四码/换新/盘点。
import {
  Entity, PrimaryGeneratedColumn, Column, ManyToOne, JoinColumn, Index, OneToMany,
} from 'typeorm';
import type { AssetStatus, QuadStatus, TagStatus, TaskStatus } from '../enums.js';
import { LegalEntity } from './org.js';
import { Address } from './geo.js';
import { Worker } from './worker.js';

@Entity('asset_batches', { comment: 'L2入库批次:公司采购行为,资产经此归属公司' })
export class AssetBatch {
  @PrimaryGeneratedColumn({ comment: '主键' })
  id!: number;

  @ManyToOne(() => LegalEntity, { nullable: false })
  @JoinColumn({ name: 'legal_entity_id' })
  legalEntity!: LegalEntity;

  @Column({ type: 'varchar', length: 32, unique: true, comment: '批次编码,如RK-202607-01' })
  code!: string;

  @Column({ type: 'varchar', length: 64, comment: '批次名称,如7月光猫批次' })
  name!: string;

  /** 批次内资产 */
  @OneToMany(() => Asset, (a) => a.batch)
  assets!: Asset[];
}

@Entity('tags', { comment: 'L2电子标签池:公司库存,预绑定后才指向资产' })
@Index(['tagNo'], { unique: true })
export class Tag {
  @PrimaryGeneratedColumn({ comment: '主键' })
  tagId!: number;

  @ManyToOne(() => LegalEntity, { nullable: false })
  @JoinColumn({ name: 'legal_entity_id' })
  legalEntity!: LegalEntity;

  @Column({ type: 'varchar', length: 32, unique: true, comment: '标签编号' })
  tagNo!: string;

  @Column({ type: 'varchar', length: 32, unique: true, comment: 'EPC码(电子标签全球码)' })
  epcCode!: string;

  @Column({ type: 'varchar', length: 8, comment: '频段,如UHF' })
  band!: string;

  @Column({ type: 'bigint', nullable: true, comment: '预绑定资产id:显式null=未绑定(禁用占位伪值),跨层回填豁免' })
  boundAssetId?: number | null;

  @Column({ type: 'varchar', length: 16, default: 'UNBOUND', comment: '状态:UNBOUND未绑定/BOUND已绑定/DISABLED禁用' })
  status!: TagStatus;

  @Column({ type: 'varchar', length: 8, comment: '电池电量,如86%' })
  battery!: string;
}

@Entity('assets', { comment: 'L4资产台账:光猫/ONU等装维物资的全生命周期' })
@Index(['assetCode'], { unique: true })
export class Asset {
  @PrimaryGeneratedColumn({ comment: '主键' })
  assetId!: number;

  @Column({ type: 'varchar', length: 32, unique: true, comment: '资产编码,如A-20260001' })
  assetCode!: string;

  @ManyToOne(() => AssetBatch, { nullable: false })
  @JoinColumn({ name: 'batch_id' })
  batch!: AssetBatch;

  @Column({ type: 'bigint', comment: '企业归属id(legal_entities)快照:入库批次所属企业' })
  legalEntityId!: number;

  @Column({ type: 'varchar', length: 128, comment: '企业名快照:改名不改历史资产' })
  legalEntityName!: string;

  @Column({ type: 'bigint', nullable: true, comment: '绑定标签id(tags表),入库未绑定时为空' })
  tagId?: number | null;

  @ManyToOne(() => Address, { nullable: true })
  @JoinColumn({ name: 'address_id' })
  address?: Address | null;

  @Column({ type: 'integer', nullable: true, comment: '经营区域id(regions表):部署地址所在区域,未部署为空,按地区统计锚点' })
  regionId?: number | null;

  @Column({ type: 'varchar', length: 64, nullable: true, comment: '经营区域名快照:改名不改历史' })
  regionName?: string | null;

  @Column({ type: 'varchar', length: 32, comment: '资产类型:光猫/ONU/路由器等' })
  type!: string;

  @Column({ type: 'varchar', length: 16, comment: '状态:IN_STOCK在库/DEPLOYED在用/MAINTENANCE维保/SCRAPPED报废' })
  status!: AssetStatus;

  /** 状态轨迹(历史:每次状态/位置变更,不随当前状态漂移) */
  @OneToMany(() => AssetLifecycle, (l) => l.asset)
  lifecycles!: AssetLifecycle[];

  /** 持有台账(历史:每次领用/部署/归还) */
  @OneToMany(() => AssetAssignment, (a) => a.asset)
  assignments!: AssetAssignment[];
}

@Entity('asset_lifecycles', { comment: 'L5资产状态轨迹:记录资产每次状态/位置变更,历史不随当前状态漂移' })
@Index(['assetId', 'changedAt'])
export class AssetLifecycle {
  @PrimaryGeneratedColumn({ comment: '主键' })
  id!: number;

  @ManyToOne(() => Asset, (a) => a.lifecycles, { nullable: false })
  @JoinColumn({ name: 'asset_id' })
  asset!: Asset;

  @Column({ type: 'varchar', length: 16, comment: '状态:IN_STOCK/DEPLOYED/MAINTENANCE/SCRAPPED' })
  status!: AssetStatus;

  @ManyToOne(() => Address, { nullable: true })
  @JoinColumn({ name: 'address_id' })
  address?: Address | null;

  @Column({ type: 'varchar', length: 64, nullable: true, comment: '事发时地址名快照:改名不改历史' })
  addressName?: string | null;

  @Column({ type: 'bigint', nullable: true, comment: '经办师傅id(部署/归还人)' })
  workerId?: number | null;

  @Column({ type: 'varchar', length: 64, nullable: true, comment: '经办师傅姓名快照:改名不改历史' })
  workerName?: string | null;

  @Column({ type: 'timestamptz', comment: '变更时间' })
  changedAt!: Date;
}

@Entity('asset_assignments', { comment: 'L4资产持有台账:记录资产每次被师傅持有/部署位置的时间段,历史归属不随当前值漂移' })
@Index(['assetId', 'effectiveFrom'])
export class AssetAssignment {
  @PrimaryGeneratedColumn({ comment: '主键' })
  id!: number;

  @ManyToOne(() => Asset, (a) => a.assignments, { nullable: false })
  @JoinColumn({ name: 'asset_id' })
  asset!: Asset;

  @ManyToOne(() => Worker, { nullable: true })
  @JoinColumn({ name: 'worker_id' })
  worker?: Worker | null;

  @Column({ type: 'varchar', length: 64, nullable: true, comment: '持有师傅姓名快照:改名不改历史台账' })
  workerName?: string | null;

  @ManyToOne(() => Address, { nullable: true })
  @JoinColumn({ name: 'address_id' })
  address?: Address | null;

  @Column({ type: 'varchar', length: 64, nullable: true, comment: '部署地址名快照:改名不改历史台账' })
  addressName?: string | null;

  @Column({ type: 'varchar', length: 128, nullable: true, comment: '变更原因,如领用/部署/归还' })
  reason?: string | null;

  @Column({ type: 'bigint', nullable: true, comment: '操作人账号id' })
  operatorAccountId?: number | null;

  @Column({ type: 'timestamptz', comment: '持有生效时间' })
  effectiveFrom!: Date;

  @Column({ type: 'timestamptz', nullable: true, comment: '持有结束时间;null=至今' })
  effectiveTo?: Date | null;
}

@Entity('quad_links', { comment: 'L5四码合一:资产-客户-端口-地址四码关联(第二码是客户,不是LOID)' })
@Index(['assetId'], { unique: true })
@Index(['customerId'], { unique: true })
@Index(['portId'], { unique: true })
@Index(['addressId'], { unique: true })
export class QuadLink {
  @PrimaryGeneratedColumn({ comment: '主键' })
  id!: number;

  @Column({ type: 'bigint', comment: '资产id(assets表)' })
  assetId!: number;

  @Column({ type: 'bigint', comment: '客户id(customers表)——口径裁定:四码第2项=客户' })
  customerId!: number;

  @Column({ type: 'bigint', comment: '端口id(ports表)' })
  portId!: number;

  @Column({ type: 'bigint', comment: '地址id(addresses表,楼栋级)' })
  addressId!: number;

  @Column({ type: 'bigint', comment: '企业归属id(legal_entities)快照:四码客户所属企业' })
  legalEntityId!: number;

  @Column({ type: 'varchar', length: 128, comment: '企业名快照:改名不改历史' })
  legalEntityName!: string;

  @Column({ type: 'varchar', length: 16, comment: '状态:LINKED一致/CONFLICT冲突/UNLINKED未关联' })
  status!: QuadStatus;
}

@Entity('replacements', { comment: 'L5换新单:故障资产换新流程' })
@Index(['replacementNo'], { unique: true })
export class Replacement {
  @PrimaryGeneratedColumn({ comment: '主键' })
  id!: number;

  @Column({ type: 'varchar', length: 32, unique: true, comment: '换新单号' })
  replacementNo!: string;

  @Column({ type: 'bigint', comment: '故障资产id' })
  assetId!: number;

  @Column({ type: 'bigint', comment: '企业归属id(legal_entities)快照:故障资产所属企业' })
  legalEntityId!: number;

  @Column({ type: 'varchar', length: 128, comment: '企业名快照:改名不改历史' })
  legalEntityName!: string;

  @Column({ type: 'varchar', length: 64, comment: '换新原因,如光猫故障' })
  reason!: string;

  @Column({ type: 'varchar', length: 8, comment: '优先级:HIGH/MEDIUM/LOW' })
  priority!: string;

  @Column({ type: 'varchar', length: 16, default: 'PENDING', comment: '状态:PENDING/DOING/DONE/FAILED' })
  status!: TaskStatus;
}

@Entity('stocktakes', { comment: 'L5盘点任务:按区域盘点资产,输出差异' })
export class Stocktake {
  @PrimaryGeneratedColumn({ comment: '主键' })
  id!: number;

  @ManyToOne(() => LegalEntity, { nullable: false })
  @JoinColumn({ name: 'legal_entity_id' })
  legalEntity!: LegalEntity;

  @Column({ type: 'varchar', length: 128, comment: '盘点范围:区域LTREE路径' })
  scope!: string;

  @Column({ type: 'smallint', comment: '进度百分比0~100' })
  progress!: number;

  @Column({ type: 'integer', comment: '差异条数(账实不符)' })
  diffCount!: number;

  @Column({ type: 'varchar', length: 16, comment: '状态:DOING进行中/DONE完成' })
  status!: string;
}
